rpc/src/lib.rs

```diff
@@ -131,34 +131,55 @@ use http::tokio_core;
 pub type HttpServer = http::Server;
 
 /// Start http server asynchronously and returns result with `Server` handle on success or an error.
-pub fn start_http<M, S, H, T, R>(
+pub fn start_http<M, S, H, T>(
 	addr: &SocketAddr,
 	cors_domains: http::DomainsValidation<http::AccessControlAllowOrigin>,
 	allowed_hosts: http::DomainsValidation<http::Host>,
 	handler: H,
 	remote: tokio_core::reactor::Remote,
 	extractor: T,
-	middleware: Option<R>,
 	threads: usize,
 ) -> ::std::io::Result<HttpServer> where
 	M: jsonrpc_core::Metadata,
 	S: jsonrpc_core::Middleware<M>,
 	H: Into<jsonrpc_core::MetaIoHandler<M, S>>,
 	T: HttpMetaExtractor<Metadata=M>,
-	R: RequestMiddleware,
 {
 	let extractor = http_common::MetaExtractor::new(extractor);
-	let mut builder = http::ServerBuilder::with_meta_extractor(handler, extractor)
+	Ok(http::ServerBuilder::with_meta_extractor(handler, extractor)
 		.threads(threads)
 		.event_loop_remote(remote)
 		.cors(cors_domains.into())
-		.allowed_hosts(allowed_hosts.into());
-
-	if let Some(dapps) = middleware {
-		builder = builder.request_middleware(dapps)
-	}
+		.allowed_hosts(allowed_hosts.into())
+		.start_http(addr)?)
+}
 
-	Ok(builder.start_http(addr)?)
+/// Same as `start_http`, but takes an additional `middleware` parameter that is introduced as a
+/// hyper middleware.
+pub fn start_http_with_middleware<M, S, H, T, R>(
```

Comment:
Not sure about the duplication here, maybe pass a function that can add something to the builder: `F: FnOnce(&mut Builder)` function instead of `Option<R>`? (Cause I beleive the reason for duplication is that in case we are passing `None` the type cannot be inferred)